This only works with `mvn quarkus:dev`. I was expecting it to work with `java -jar` and this is the output:

```
/tmp/greeting-quarkus ᐅ java -jar target/greeting-quarkus-0.1.0-runner.jar
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-10-13 13:41:18,229 ERROR [io.qua.run.Application] (main) Failed to start application (with profile prod): java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:609)
	at io.quarkus.runtime.Application.start(Application.java:90)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:93)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:62)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:104)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
Caused by: java.lang.RuntimeException: Unable to start HTTP server
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:501)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServer(VertxHttpRecorder.java:224)
	at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket-2064782366.deploy_0(VertxHttpProcessor$openSocket-2064782366.zig:136)
	at io.quarkus.deployment.steps.VertxHttpProcessor$openSocket-2064782366.deploy(VertxHttpProcessor$openSocket-2064782366.zig:40)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:543)
	... 6 more
Caused by: java.util.concurrent.ExecutionException: java.net.BindException: Address already in use
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:467)
	... 10 more
Caused by: java.net.BindException: Address already in use
	at java.base/sun.nio.ch.Net.bind0(Native Method)
	at java.base/sun.nio.ch.Net.bind(Net.java:455)
	at java.base/sun.nio.ch.Net.bind(Net.java:447)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:227)
	at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:550)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
	at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
	at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
	at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:248)
	at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:356)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasksFrom(SingleThreadEventExecutor.java:428)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:377)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:488)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/dev/IsolatedDevModeMain.java
@@ -8,6 +8,7 @@ import java.io.Closeable;
 import java.io.IOException;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
+import java.net.BindException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
@@ -108,35 +109,40 @@ public class IsolatedDevModeMain implements BiConsumer<CuratedApplication, Map<S
                 runner = start.runMainClass(context.getArgs());
                 firstStartCompleted = true;
             } catch (Throwable t) {
-                deploymentProblem = t;
-                if (!augmentDone) {
-                    log.error("Failed to start quarkus", t);
+                Throwable rootCause = t;
+                while (rootCause.getCause() != null) {
+                    rootCause = rootCause.getCause();
                 }
-                if (!context.isAbortOnFailedStart()) {
-                    //we need to set this here, while we still have the correct TCCL
-                    //this is so the config is still valid, and we can read HTTP config from application.properties
-                    log.info("Attempting to start hot replacement endpoint to recover from previous Quarkus startup failure");
-                    if (RuntimeUpdatesProcessor.INSTANCE != null) {
-                        Thread.currentThread().setContextClassLoader(curatedApplication.getBaseRuntimeClassLoader());
-
-                        try {
-                            if (!InitialConfigurator.DELAYED_HANDLER.isActivated()) {
-                                Class<?> cl = Thread.currentThread().getContextClassLoader()
-                                        .loadClass(LoggingSetupRecorder.class.getName());
-                                cl.getMethod("handleFailedStart").invoke(null);
+                if (!(rootCause instanceof BindException)) {
+                    deploymentProblem = t;
+                    if (!augmentDone) {
+                        log.error("Failed to start quarkus", t);
+                    }
+                    if (!context.isAbortOnFailedStart()) {
+                        //we need to set this here, while we still have the correct TCCL
+                        //this is so the config is still valid, and we can read HTTP config from application.properties
+                        log.info(
+                                "Attempting to start hot replacement endpoint to recover from previous Quarkus startup failure");
+                        if (RuntimeUpdatesProcessor.INSTANCE != null) {
+                            Thread.currentThread().setContextClassLoader(curatedApplication.getBaseRuntimeClassLoader());
+                            try {
+                                if (!InitialConfigurator.DELAYED_HANDLER.isActivated()) {
+                                    Class<?> cl = Thread.currentThread().getContextClassLoader()
+                                            .loadClass(LoggingSetupRecorder.class.getName());
+                                    cl.getMethod("handleFailedStart").invoke(null);
+                                }
+                                RuntimeUpdatesProcessor.INSTANCE.startupFailed();
+                            } catch (Exception e) {
+                                close();
+                                log.error("Failed to recover after failed start", e);
+                                //this is the end of the road, we just exit
+                                //generally we only hit this if something is already listening on the HTTP port
+                                //or the system config is so broken we can't start HTTP
+                                System.exit(1);
                             }
-                            RuntimeUpdatesProcessor.INSTANCE.startupFailed();
-                        } catch (Exception e) {
-                            close();
-                            log.error("Failed to recover after failed start", e);
-                            //this is the end of the road, we just exit
-                            //generally we only hit this if something is already listening on the HTTP port
-                            //or the system config is so broken we can't start HTTP
-                            System.exit(1);
                         }
                     }
                 }
-
             }
         } finally {
             Thread.currentThread().setContextClassLoader(old);
@@ -176,9 +182,15 @@ public class IsolatedDevModeMain implements BiConsumer<CuratedApplication, Map<S
                 firstStartCompleted = true;
             } catch (Throwable t) {
                 deploymentProblem = t;
-                log.error("Failed to start quarkus", t);
-                Thread.currentThread().setContextClassLoader(curatedApplication.getAugmentClassLoader());
-                LoggingSetupRecorder.handleFailedStart();
+                Throwable rootCause = t;
+                while (rootCause.getCause() != null) {
+                    rootCause = rootCause.getCause();
+                }
+                if (!(rootCause instanceof BindException)) {
+                    log.error("Failed to start quarkus", t);
+                    Thread.currentThread().setContextClassLoader(curatedApplication.getAugmentClassLoader());
+                    LoggingSetupRecorder.handleFailedStart();
+                }
             }
         } finally {
             restarting = false;

--- a/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
@@ -1,5 +1,6 @@
 package io.quarkus.runtime;
 
+import java.net.BindException;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.Set;
@@ -14,6 +15,7 @@ import javax.enterprise.inject.spi.Bean;
 import javax.enterprise.inject.spi.BeanManager;
 import javax.enterprise.inject.spi.CDI;
 
+import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
 import org.graalvm.nativeimage.ImageInfo;
 import org.jboss.logging.Logger;
 import org.wildfly.common.lock.Locks;
@@ -139,8 +141,28 @@ public class ApplicationLifecycleManager {
             }
         } catch (Exception e) {
             if (exitCodeHandler == null) {
-                Logger.getLogger(Application.class).errorv(e, "Failed to start application (with profile {0})",
-                        ProfileManager.getActiveProfile());
+                Throwable rootCause = e;
+                while (rootCause.getCause() != null) {
+                    rootCause = rootCause.getCause();
+                }
+                Logger applicationLogger = Logger.getLogger(Application.class);
+                if (rootCause instanceof BindException) {
+                    int port = ConfigProviderResolver.instance().getConfig()
+                            .getOptionalValue("quarkus.http.port", Integer.class).orElse(8080);
+                    applicationLogger.error("Port " + port + " seems to be in use by another process. " +
+                            "Quarkus may already be running or the port is used by another application.");
+                    if (System.getProperty("os.name").startsWith("Windows")) {
+                        applicationLogger.info("Use 'netstat -a -b -n -o' to identify the process occupying the port.");
+                        applicationLogger.info("You can try to kill it with 'taskkill /PID <pid>' or via the Task Manager.");
+                    } else {
+                        applicationLogger
+                                .info("Use 'netstat -anop | grep " + port + "' to identify the process occupying the port.");
+                        applicationLogger.info("You can try to kill it with 'kill -9 <pid>'.");
+                    }
+                } else {
+                    applicationLogger.errorv(rootCause, "Failed to start application (with profile {0})",
+                            ProfileManager.getActiveProfile());
+                }
             }
             stateLock.lock();
             try {

